apps/prairielearn/src/lib/bot-grading.ts

```diff
@@ -0,0 +1,182 @@
+  assessment_question,
+  urlPrefix,
+}: {
+  question: Question;
+  course: Course;
+  course_instance_id?: string;
+  assessment_question: AssessmentQuestion;
+  urlPrefix: string;
+}): Promise<string> {
+  // if OpenAI API Key and Organization are not provided, throw error
+  if (!config.openAiApiKey || !config.openAiOrganization) {
+    throw new error.HttpStatusError(501, 'Not implemented (feature not available)');
+  }
+  const openaiconfig = { apiKey: config.openAiApiKey, organization: config.openAiOrganization };
+
+  const question_course = await getQuestionCourse(question, course);
+
+  const serverJob = await createServerJob({
+    courseId: course.id,
+    courseInstanceId: course_instance_id,
+    assessmentId: assessment_question.assessment_id,
+    authnUserId: '1',
+    type: 'botGrading',
+    description: 'Use LLM to grade assessment question',
+  });
+
+  serverJob.executeInBackground(async (job) => {
+    const openai = new OpenAI(openaiconfig);
+    job.info('OpenAI client ready! ');
+
+    // get all instance questions
+    const result = await queryRows(
+      sql.select_instance_questions_manual_grading,
+      {
+        assessment_id: assessment_question.assessment_id,
+        assessment_question_id: assessment_question.id,
+      },
+      InstanceQuestionSchema,
+    );
+
+    job.info(`Found ${result.length} submissions to grade! `);
+
+    let error_count = 0;
+    let output_count = 0;
+    let output: string | null = null;
+
+    // get each instance question
+    for (const instance_question of result) {
+      const { variant, submission } = await queryRow(
+        sql.select_last_variant_and_submission,
+        { instance_question_id: instance_question.id },
+        SubmissionVariantSchema,
+      );
+
+      // build urls for the question server
+      const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question);
+
+      // get question html
+      const questionModule = questionServers.getModule(question.type);
+      const { courseIssues, data } = await questionModule.render(
+        { question: true, submissions: false, answer: false },
+        variant,
+        question,
+        null,
+        [],
+        question_course,
+        urls,
+      );
+      if (courseIssues.length) {
+        job.info(courseIssues.toString());
+        job.error('Error occurred');
+        job.fail('Errors occurred while bot grading, see output for details');
+      }
+      const question_prompt_raw = data.questionHtml;
+      const $ = cheerio.load(question_prompt_raw, null, false);
+      $('script').remove();
+      const question_prompt = $.html();
+
+      const render_submission_results = await questionModule.render(
+        { question: false, submissions: true, answer: false },
+        variant,
+        question,
+        submission,
+        [submission],
+        question_course,
+        urls,
+      );
+      const student_answer = render_submission_results.data.submissionHtmls[0];
```

Comment:
Is there a reason you strip the scripts from the prompt but not the student answer?

Reply:
This might be related to my unfamiliarity with html, but I did not see any script tags when I print out the student answer?

Reply:
I guess in the examples that we were looking at, there were no scripts in the submission panel so we didn't think about it, but I supposed in general they could be there, so we should get rid of them there too @Victorsss-Orz